src/docs/tree_widget.md

```diff
@@ -0,0 +1,723 @@
+
+### Expandable/Collapsible Tree Nodes
+
+The management of expandable/collapsible tree nodes is already built into the framework. The only thing we need to change is to make our `ExampleTreeNode` interface conform to `ExpandableTreeNode` instead of `CompositeTreeNode` and initialize the `expanded` property to `false` in the `TreeViewExampleTreeItemFactory` class.
+
+### Lazy Child Resolution
+
+Usually, we don't want to initialize the complete tree upfront. Instead, only the visible items, namely the children of the root node, should be initialized. All the other children should be initialized when they are needed, which is essentially when their containing parent node is expanded.
+
+So, in the `TreeViewExampleModel.init()` method, we change the initialization to
+
+```typescript
+        ...
+        EXAMPLE_DATA.map(item => this.itemFactory.toTreeNode(item))
+            .forEach(node => CompositeTreeNode.addChild(root, node));
+        this.tree.root = root;
+````
+
+and implement the dynamic resolution of child nodes by subclassing `TreeImpl`:
+
+```typescript
+export class TreeviewExampleTree extends TreeImpl {
+    @inject(TreeViewExampleTreeItemFactory) private readonly itemFactory: TreeViewExampleTreeItemFactory;
+
+    override async resolveChildren(parent: CompositeTreeNode): Promise<TreeNode[]> {
+        // root children are initialized once and never change
+        if (parent.id === ROOT_NODE_ID) {
+            return [...parent.children];
+        }
+
+        // non-container classes do not have children
+        if (!ExampleTreeNode.is(parent)) {
+            return [];
+        }
+
+        // (optional) caching: if the children are resolved and no child was added/removed, reuse 
+        if (parent.children.length === parent.data.children?.length) {
+            return [...parent.children];
+        }
+
+        // simulate asynchronous loading of children
+        await wait(2000);
+        return (parent.data.children ?? []).map(i => this.itemFactory.toTreeNode(i));        
+    }
+}
+```
+
+Under the hood, the tree expansion is handled by the `TreeExpansionServiceImpl` implementation which maintains the `expanded` state of `ExpandableTreeNode`, sends events related to expanding and collapsing subtrees, and also calls `Tree.refresh()` for the expanded node. The `refresh()` method in turn calls `resolveChildren()` which can be implemented to asyncronously provide the child nodes for the given parent. As a bonus, `Tree.refresh()` marks the node as `busy` until the promise is resolved, leading to a nice busy marker in the form of a spinning circle if the promise is not resolved within a certain amount of time (800ms). This can be observed in this demo code due to the `wait(2000)` call.
```

Comment:
`TreeExpansionServiceImpl` implementation -> `TreeExpansionService` implementation (could be anything that is bound to that service)